### PR TITLE
Fix #6975 Reproducible bean generation

### LINKED ISSUE #6975: Reproducible bean generation
The ordering of the bytecode generated by Arc is not determininstic. A PR follows.

### PR DESCRIPTION
**Pull request by ppalaga**
Blocked by https://github.com/wildfly/jandex/issues/71

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java [hunk 1]
```diff
@@ -419,7 +420,14 @@ void init(List<Throwable> errors, Consumer<BytecodeTransformer> bytecodeTransfor
         if (disposer != null) {
             disposer.init(errors);
         }
-        interceptedMethods.putAll(initInterceptedMethods(errors, bytecodeTransformerConsumer, removeFinalForProxyableMethods));
+
+        synchronized (lock) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: Why the lock, the volatile field and all the magic here?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 2]
```diff
@@ -203,8 +204,12 @@ public ContextConfigurator configure(Class<? extends Annotation> scopeAnnotation
 
     BeanRegistrar.RegistrationContext registerBeans(List<BeanRegistrar> beanRegistrars) {
         List<InjectionPointInfo> injectionPoints = new ArrayList<>();
-        this.beans.addAll(findBeans(initBeanDefiningAnnotations(beanDefiningAnnotations, stereotypes.keySet()), observers,
-                injectionPoints, jtaCapabilities));
+        final List<BeanInfo> addedBeans = findBeans(initBeanDefiningAnnotations(beanDefiningAnnotations, stereotypes.keySet()),
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: What's the reason for this?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 3]
```diff
@@ -562,12 +567,12 @@ static ScopeInfo getValidScope(Collection<ScopeInfo> stereotypeScopes, Annotatio
     private List<BeanInfo> findBeans(Collection<DotName> beanDefiningAnnotations, List<ObserverInfo> observers,
             List<InjectionPointInfo> injectionPoints, boolean jtaCapabilities) {
 
-        Set<ClassInfo> beanClasses = new HashSet<>();
-        Set<MethodInfo> producerMethods = new HashSet<>();
-        Set<MethodInfo> disposerMethods = new HashSet<>();
-        Set<FieldInfo> producerFields = new HashSet<>();
-        Map<MethodInfo, Set<ClassInfo>> syncObserverMethods = new HashMap<>();
-        Map<MethodInfo, Set<ClassInfo>> asyncObserverMethods = new HashMap<>();
+        Set<ClassInfo> beanClasses = new TreeSet<>();
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Why sometimes `TreeSet` and sometimes `LinkedHashSet`?

**Review by mkouba — COMMENTED** [review]
I can't say I like the idea of replacing the `Set` and `Map` implementations like this, ie. without obvious meaning. Anyway, I think we should also (1) verify that it does not hurt the build performance, (2) document the replacements and hacks properly.

**Comment by ppalaga** [comment]
The last iteration of https://github.com/quarkusio/quarkus/pull/6972 may eliminate the need for some `Tree[Set|Map]`s. Let me revisit.

**Comment by ppalaga** [comment]
Yes, the sorted collections are not needed with the last iteration of https://github.com/quarkusio/quarkus/pull/6972 .

This PR is not carved in stone. I admit that I am generally ignorant about Arc internals and that this PR is party a blind replace of non-repeatable iteration order collections with the repeatable ones. It is quite likely that not all the replacements are needed (or other solutions are possible), but because they are so many, I was not able to check them one by one. 

However, the current state is a good baseline, because it helps to make one basic build scenario (the hibernate-validator itest) binary reproducible. Unfortunately, not all changes needed to reach that are public yet. I need to prepare one or two more PRs.

So how can we proceed here? I am thankful for the initial feedback. Let me go through the individual comments and let me push the version without `Tree[Set|Map]`. The discussion will make much more sense once I send the missing PRs so that reviewers can also clearly see the impact of the changes on the bytecode level.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java (hunk 1 above)
Comment (ppalaga) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: `interceptedMethods` being originally a `ConcurrentHashMap` suggested the possibility that `init()` might be called multiple times from concurrent threads. So I over protected it to see if it is going to fail somewhere. I am open to simplify it.

Comment (ppalaga) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: BTW `interceptedMethods` needs to be a LinkedHashMap for the sake of reproducibility.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 2 above)
Comment (ppalaga) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: It is a leftover after some debug output that was there to check whether the order stays stable during repeated builds. Let me revert it.

**Comment by ppalaga** [comment]
f8b3805247e0c5b24e1f2d1ef29fdfdf7046a09e : 

* uses only Linked collections
* cleans a leftover

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 3 above)
Comment (ppalaga) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Only linked collections are used in f8b3805247e0c5b24e1f2d1ef29fdfdf7046a09e

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 2 above)
Comment (ppalaga) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Reverted in f8b3805247e0c5b24e1f2d1ef29fdfdf7046a09e

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java]: `init()` is only called once. I'm not sure about the `ConcurrentHashMap` though - I'll verify this when I have some spare time.

**Comment by ppalaga** [comment]
As agreed on the developer mailing list, Quarkus build reproducibility is not required at this time, see https://groups.google.com/forum/#!topic/quarkus-dev/6geWtb5HOdc Thus closing.

In case this becomes a topic in the future, here is my latest code https://github.com/ppalaga/quarkus/commits/native-image-cache-and-build-reproducibility

**Pull request closed without merging**